File: app/src/main/java/com/example/david/lists/view/userlistlist/IUserListViewContract.java
package com.example.david.lists.view.userlistlist;

import com.example.david.lists.data.datamodel.UserList;
import com.example.david.lists.view.authentication.IAuthContract;

import java.util.List;

public interface IUserListViewContract {
    interface View {
        void openUserList(UserList userList);

        void confirmSignOut();

        void openAuthentication(IAuthContract.AuthGoal authGoal, int requestCode, String intentExtraAuthResultKey);

        void openAddDialog(int position);

        void openEditDialog(UserList userList);

        void submitList(List<UserList> viewData);

        void notifyUserOfDeletion(String message);

        void setStateDisplayList();

        void setStateLoading();

        void setStateError(String message);

        void recreateView();
    }

    interface Logic {
        void onStart();

        void userListSelected(int position);

        void add();

        void edit(UserList userList);

        void dragging(int fromPosition, int toPosition, IUserListViewContract.Adapter adapter);

        void movedPermanently(int newPosition);

        void delete(int position, IUserListViewContract.Adapter adapter);

        void undoRecentDeletion(Adapter adapter);

        void deletionNotificationTimedOut();

        void signOut();

        void signOutConfirmed();

        void signIn();

        void authResult(IAuthContract.AuthResult authResult);

        void setNightMode(boolean isMenuItemChecked);

        boolean isUserAnon();

        boolean isNightModeEnabled();

        void onDestroy();
    }

    interface ViewModel {
        void setViewData(List<UserList> userLists);

        List<UserList> getViewData();

        void setTempPosition(int position);

        List<UserList> getTempList();

        int getTempPosition();

        int getRequestCode();

        String getIntentExtraAuthResultKey();

        String getMsgInvalidUndo();

        String getMsgDeletion();

        String getErrorMsg();

        String getErrorMsgEmptyList();
    }

    interface Adapter {
        void submitList(List<UserList> list);

        void move(int fromPosition, int toPosition);

        void remove(int position);

        void reAdd(int position, UserList userList);
    }
}

File: app/src/main/java/com/example/david/lists/view/userlistlist/IUserListViewContract.kt
package com.example.david.lists.view.userlistlist

import com.example.domain.datamodel.UserList

interface IUserListViewContract {
    interface View {
        fun openUserList(userList: UserList)

        fun openPreferences()

        fun openAddDialog(position: Int)

        fun openEditDialog(userList: UserList)

        fun setViewData(viewData: List<UserList>)

        fun notifyUserOfDeletion(message: String)

        fun setStateDisplayList()

        fun setStateLoading()

        fun setStateError(message: String)

        fun showMessage(message: String)
    }

    interface Logic {
        val isNightModeEnabled: Boolean

        fun onStart()

        fun userListSelected(position: Int)

        fun add()

        fun edit(position: Int)

        fun dragging(fromPosition: Int, toPosition: Int, adapter: Adapter)

        fun movedPermanently(newPosition: Int)

        fun delete(position: Int, adapter: Adapter)

        fun undoRecentDeletion(adapter: Adapter)

        fun deletionNotificationTimedOut()

        fun preferencesSelected()

        fun setNightMode(isMenuItemChecked: Boolean)

        fun onDestroy()
    }

    interface ViewModel {
        var viewData: MutableList<UserList>

        val tempList: MutableList<UserList>

        var tempPosition: Int

        val msgDeletion: String

        val errorMsg: String

        val errorMsgEmptyList: String

        val errorMsgInvalidUndo: String
    }

    interface Adapter {
        fun setData(list: List<UserList>)

        fun move(fromPosition: Int, toPosition: Int)

        fun remove(position: Int)

        fun reAdd(position: Int, userList: UserList)
    }
}
